core/trino-main/src/main/java/io/trino/sql/SessionSpecificationEvaluator.java

```diff
@@ -0,0 +1,208 @@
+    {
+        ResolvedSessionSpecifications resolvedSessionSpecifications = resolve(session, parameters, specifications);
+        return overrideProperties(session, resolvedSessionSpecifications);
+    }
+
+    public ResolvedSessionSpecifications resolve(Session session, Map<NodeRef<Parameter>, Expression> parameters, List<SessionSpecification> specifications)
+    {
+        ImmutableMap.Builder<String, String> sessionProperties = ImmutableMap.builder();
+        Table<String, String, String> catalogProperties = HashBasedTable.create();
+        Set<QualifiedName> seenPropertyNames = new HashSet<>();
+
+        for (SessionSpecification specification : specifications) {
+            List<String> nameParts = specification.getName().getParts();
+
+            if (!seenPropertyNames.add(specification.getName())) {
+                throw semanticException(INVALID_SESSION_PROPERTY, specification, "Session property %s already set", specification.getName());
+            }
+
+            if (nameParts.size() == 1) {
+                Optional<PropertyMetadata<?>> systemSessionPropertyMetadata = sessionPropertyManager.getSystemSessionPropertyMetadata(nameParts.getFirst());
+                if (systemSessionPropertyMetadata.isEmpty()) {
+                    throw semanticException(INVALID_SESSION_PROPERTY, specification, "Session property %s does not exist", specification.getName());
+                }
+                sessionProperties.put(nameParts.getFirst(), toSessionValue(session, parameters, specification, systemSessionPropertyMetadata.get()));
+            }
+
+            if (nameParts.size() == 2) {
+                CatalogHandle catalogHandle = getRequiredCatalogHandle(plannerContext.getMetadata(), session, specification, nameParts.getFirst());
+                Optional<PropertyMetadata<?>> connectorSessionPropertyMetadata = sessionPropertyManager.getConnectorSessionPropertyMetadata(catalogHandle, nameParts.getLast());
+                if (connectorSessionPropertyMetadata.isEmpty()) {
+                    throw semanticException(INVALID_SESSION_PROPERTY, specification, "Session property %s does not exist", specification.getName());
+                }
+                catalogProperties.put(nameParts.get(0), nameParts.get(1), toSessionValue(session, parameters, specification, connectorSessionPropertyMetadata.get()));
+            }
+        }
+
+        return new ResolvedSessionSpecifications(sessionProperties.buildOrThrow(), catalogProperties.rowMap());
+    }
+
+    public Session overrideProperties(Session session, ResolvedSessionSpecifications resolvedSessionSpecifications)
+    {
+        requireNonNull(resolvedSessionSpecifications, "resolvedSessionSpecifications is null");
+
+        validateSystemProperties(session, resolvedSessionSpecifications.systemProperties());
+
+        // Catalog session properties were already evaluated so we need to evaluate overrides
+        if (session.getTransactionId().isPresent()) {
```

Comment:
did you consider moving validation logic to `Session` class where it already is? You can put it in `Session.withProperties` maybe?

Reply:
It requires some changes for injecting `AccessControl` to `Session`. Added a TODO comment.